### PR TITLE
Opencv bridge update

### PR DESCRIPTION
**Pull request by PierreWargnier**
BUG: Fix Mat to Image conversions for OpenCV 4.x; add support for int32

This fixes issue https://github.com/InsightSoftwareConsortium/ITK/issues/1651 in a lasting manner by introducing specific implementations for functions CVMatToITKImage and ITKImageToCVMat that no longer depend on the legacy C API while avoiding unnescessary memory copies.

This also fixes issue https://github.com/InsightSoftwareConsortium/ITK/issues/1906 by adding support for int pixel type (IPL_DEPTH_32S in the legacy C API, CV_32S in the C++ API) for both OpenCV legacy C and C++ APIs.

In addition, exceptions for pixel/component type mismatches and impossible conversions were added to both IplImageToITKImage and CVMatToITKImage functions. Some extra functions had to be introduced to ensure minimal code duplication/redundancy.

Desired behavior was discussed with @dzenanz . It will remain the same as before, only with extra checks for type mismatches and unsupported color conversions.

Note: for future improvements it could be useful to allow 2, 3 and 4-channels conversions for any OpenCV-supported types (uchar, char, ushort, short, int, float, double). The only constraints are that color conversions will only be done when supported and in other cases, a simple buffer copy should be done. It will then be up to the user to manage data representation. Support for RGBA/BGRA would also be nice, but I am not sure if it is supported in ITK.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/Video/BridgeOpenCV/include/itkOpenCVImageBridge.hxx [hunk 1]
```diff
@@ -251,14 +297,135 @@ template <typename TInputImageType>
...
+  //
+  // Set up the output image
+  //
+  Mat          tmp;
+  unsigned int w = static_cast<unsigned int>(size[0]);
+  unsigned int h = static_cast<unsigned int>(size[1]);
+
+  //
+  // set the depth correctly based on input pixel type
+  //
```
Comment (dzenanz) [inline Modules/Video/BridgeOpenCV/include/itkOpenCVImageBridge.hxx:350]: These empty comment lines before  and after the actual comment are not exactly needed. They reduce code density necessitating more scrolling. Please remove them.

Comment (PierreWargnier) [reply Modules/Video/BridgeOpenCV/include/itkOpenCVImageBridge.hxx:350]: I actually was not the one to format the comments like that. It was in the original code so I thought I should leave it like that. I do agree however that this is not a great coding style. I also realised that I left one include (<iostream>) that was only for debug when I was testing my changes.

I am willing to  make these corrections. However, this is the first time I submit a pull request and I am not sure how to submit the changes.

If I understand correctly, you also want me to ammend my commit message to include the comment about ideas for further improvements. Is that correct?

Comment (dzenanz) [reply Modules/Video/BridgeOpenCV/include/itkOpenCVImageBridge.hxx:350]: Corrections: amend the commit and then force-push to the same branch (`opencv-bridge-update`).

Ideas for future improvement were my message to you, and should not be included in commit message. Thanks for working on this!

Comment (PierreWargnier) [reply Modules/Video/BridgeOpenCV/include/itkOpenCVImageBridge.hxx]: Hi @dzenanz, the cleanup and correction of comment format is done. I also removed the useless comments saying "return result".

**Comment by PierreWargnier** [comment]
Regarding the additional support of arbitrary number of channels, it would be nice to have but I do not think it is essential. I personnally do not need it at the moment and do not really have time to implement it right now.

**Pull request merged**